TechnicalDocuments/Proposals/BabylonDesignLibrary.md

```diff
@@ -0,0 +1,208 @@
+# Babylon Design Library
+
+* Author(s): Giorgos Tsiapaliokas
+* Review Manager: Anders Ha
+
+## Introduction
+This proposal is about introducing a Design Library.
+
+## Motivation
+With our current infrastrure we have have reusable components which don't reuse the same style,
+we provide different styesheets based on the design.
+
+The idea behind this is to have reusable components with a predefined stylesheet displaying different data.
+The benefit of having a design library is
+1. to develop new UIs faster.
+2. to catch inconsistencies in the UI due to a design flaw or due to an engineering mistake.
+
+The requirement is to cover most of the common use cases and *not* to extend to edge cases.
+Also we must reuse our current tools like `Bento`.
+
+For instance, in our codebase a common UI element is a rounded button with background color. 
+For that we use `Component.Button` and a stylesheet while this approach is functional is has a few drawbacks.
+The attributes of the button (background color, alpha, corner radius etc) can either 
+1. have been redefined in another part of the codebase which is good from a design point of view *and if* we search
+for them we can use them, *if we don't* search for them we will just reimplement something which already exists.
+2. or the attributes of the button haven't been redefined in another part of the application and we may not
+notice that so a design flaw will be implemented. We are humans after all :).
+
+
+## Proposed solution
+We can introduce `DesignLibrary` which will be responsible for creating these components following
+the semantics of the design styleguide which will be provided to us.
+
+The purpose of  `DesignLibrary` isn't to expose the full potential of `Bento` components but to 
+follow specific design guidelines.
+As an example of this is: 
+A `Bento` component supports different behaviours like  `Deletable` the goal of this
+proposal isn't to replicate them or to make a wrapper around them but to follow specific design guidelines
+and to offer reusable UI elements.
+
+Also with `DesignLibrary` we want to limit the exposure of specific components, in order to be able to introduce  
+drastic design changes in the future as easy as possible.
+For example `DesignLibrary.ComponentsBuilder.button` returns `AnyRenderable` instead of `Component.Button`
+because in the future we may decide that we need to use a different button component. We want to depend on abstractions.
+
+Furthermore, `DesignLibrary` can also host alerts, custom viewcontroller animations, etc.
+
+``` swift
+
+struct DesignLibrary {
+    let tokens: Tokens
+
+    var viewControllerBuilder: ViewControllerBuilder {
```

Comment:
As mentioned in the call, IMO we probably shouldn't vend this as an instance property of `DesignLibrary`, which represents an a particular snapshot of the design library parameters. The relationship would be kinda weird, in the sense that we would have view controllers produced from a snapshot, that could set to observe the `Property<DesignLibrary>` in `Current`.

Particularly:
1. Bento modals being made would be observing `Current.designLibrary` updates itself, so the builder method is essentially a factory of `(Property<DesignLibrary>) -> UIViewController`.
2. Other non-Bento modals can pull out the current design library & get the necessary parameters from there, so the builder method is essentially a factory of `(Property<DesignLibrary>) -> DesignLibrary >>> (DesignLibrary) -> UIViewController`.

We may keep the modal builders namespaced under `DesignLibrary` for discoverability purpose. But `ViewControllerBuilder` should use `Current.designLibrary` or take `Property<DesignLibrary>`, and be treated like any other screen builds.